Title Screen Redesign

babaissarkar opened the pull request:
Resolves #4531 
Related issue #8951

Redesign of the Title screen based on @Dalas121 's proposal https://github.com/wesnoth/wesnoth/issues/4531#issuecomment-2133921535.

Note: This PR's scope is limited to the title screen only. If I decide to redo the UI (the various dialogs and controls), that will be done in another PR.

Work so far :

![Screenshot from 2024-07-16 09-18-34](https://github.com/user-attachments/assets/bdc55834-c987-4749-854b-f4ba3377a54b)

Dalas121 commented:
Wow, really impressive stuff! I honestly didn't think this was possible with the current GUI code, so I'm amazed you've managed it.

Having a big "Load" button makes a lot of sense, especially since we're moving away from a standalone tutorial.  And yeah, @stevecotton  makes a good point in #4531 that "Tutorial" probably isn't necessary, so a good thing you removed it.

I feel that the most recent "side panel" screenshots posted in #4531 possibly feel a bit more modern/stylish than the standalone buttons from my older mockup (used here). A side panel would also replacing our outdated "popup-style" windows with "menu-style", like in this image from the game Northguard:

![northgard-credits-500x281](https://github.com/wesnoth/wesnoth/assets/33790750/59503827-969e-448f-8a2a-25a7cbaca044)

But ultimately it's a matter of opinion, and I have no idea how much work it would be to try out - would like to hear others' thoughts.

CelticMinstrel commented:
I think the version number needs an outline for contrast.

babaissarkar commented:
@Dalas121 I'm quite happy with with your original design, and I think the Northgard screenshot's menu design looks too much sci-fi and not enough fantasy to me. If possible we would want to have wesnoth's design as much unique as possible (copyright issues). I'm even terrified to copying their color schemes, though it mightn't hurt to have some inspriation here and there.

Dalas121 commented:
> I'm quite happy with with your original design

I put together the revision after getting a couple comments that the original design still looked outdated, but ultimately it's your call.

If you do want to stick with a centered menu, I suggest moving away from having separate buttons like I included in my original mockup.  To the best of my knowledge visually distinct buttons aren't really as much of a thing as they used to be - AFAIK clickable words are the new standard.

> If possible we would want to have wesnoth's design as much unique as possible (copyright issues). I'm even terrified to copying their color schemes

I absolutely agree that we don't want to risk copyright, but both the left-side panel and current centered design are already very common design patterns.  Given how many games there are and how few different main menu layouts are used, I think it's going to be almost impossible to come up with something completely unique that's also familiar and intuitive for players.  

I'm involved with web design at work, and one of the core principles is that pretty much every web UI problem has already been "solved".  There's a global standard, and everybody kinda uses that same design - makes it simple for customers.

Not trying to directly copy Northguard (I've never even played it; it's just one example among many that came up when I was looking for modern UIs), but I do think this is a case where we want to find the industry standard and follow that.

babaissarkar commented:
4c3c54ae142425aa1856d192854eb3187c3e6420 allows backgrounds to be randomly selected at launch from the list here :
https://github.com/wesnoth/wesnoth/blob/4c3c54ae142425aa1856d192854eb3187c3e6420/data/game_config.cfg#L64

ForestDragon-wesnoth commented:
To be honest, I don't think this style fits wesnoth all that well. I very much prefer the old UI design style over this.

ForestDragon-wesnoth commented:
I think this in general is one of the many symptons of a larger problem - chasing after progress for the sake of progress devs forget the fact that nostalgia is the main factor keeping the community active.

stevecotton commented:
> To the best of my knowledge visually distinct buttons aren't really as much of a thing as they used to be - AFAIK clickable words are the new standard.

The rest of the game's UI uses clickable buttons, so that would be an inconsistency. If you want to use clickable words on the title screen, then I think you've expanded the task from "redesign the title screen" to "redesign the whole UI".

Dalas121 commented:
> devs forget the fact that nostalgia is the main factor keeping the community active.

Unfortunately, nostalgia can only keep a game on life support, never bring in new players. If we don't at least try to keep a healthy player count, I worry that Wesnoth will continue withering away as it has been.

IMO the biggest potential turn-offs for new players are UI, the tutorial/first campaign, and RNG - and thus I feel that those are probably the most important issues we should aim to address.

> The rest of the game's UI uses clickable buttons, so that would be an inconsistency. If you want to use clickable words on the title screen, then I think you've expanded the task from "redesign the title screen" to "redesign the whole UI".

Fair enough.  If web design applied to Wesnoth that's absolutely something I'd love to try and work on, but I have neither the skill nor time to learn our UI tools.  All I can do to help here is make mockups.

If we do want to reduce the scope of potential UI changes - both to keep it doable and to avoid alienating veteran players - perhaps we could focus on:
- keeping the OG main menu buttons, but de-emphasizing tips and swapping the background map for game art (with a setting to revert it for veterans)
- improving the campaign selection menu to better show how timelines, arcs, characters, etc link together, and to highlight the tutorial/beginner campaign
- revamping the in-game UI. Out of all our UI elements, IMO the in-game looks the most outdated. I've personally watched several friends lose interest in Wesnoth the instant they saw in-game images. Part of that is probably the terrain graphics, which I don't expect we have the desire or effort to change anytime soon, but at least part of that is the in-game UI.

While we're on the topic, I posted some in-game UI mockups some time ago on the discord. Colors and styles would need changes to better match with the existing menus UI, if we're planning to keep that.

![ingame-ui-closed](https://github.com/wesnoth/wesnoth/assets/33790750/c53e09c2-0a44-4ca5-9aa8-e88ac90dc1ae)
![ingame-ui-open](https://github.com/wesnoth/wesnoth/assets/33790750/918b1336-b9e5-4713-89ba-630cbbd76198)

babaissarkar commented:
I'd like to add a few more things from my perspective:
> IMO the biggest potential turn-offs for new players are UI, the tutorial/first campaign, and RNG - and thus I feel that those are probably the most important issues we should aim to address.
* I'd add lack of dedicated smartphone versions (Android and iOS) to the list. The number of people who only play on a phone are fairly large.

* I'd like to remind that GUI2 is fully skinnable (that was one of the main reasons for switching to GUI2 in the first place). So it is not very hard to have a setting in the Preferences that allows one to switch between multiple themes. In fact, it should be possible for Add-ons to have their own themes as well. (Sadly, nobody managed to add that yet, but I do have it on my TODO list). We can ship Wesnoth with two themes: those who prefer the previous look could simply choose that theme, while those who prefer this one can select this. There can even be others. GIMP for example ships with 5 themes, LibreOffice with multiple themes as well. Though I can't seem to recall a game that does have multiple themes, I believe Wesnoth can fairly easily manage that, since moddablity is one of its core strengths.

ForestDragon-wesnoth commented:
>     * I'd add lack of dedicated smartphone versions (Android and iOS) to the list. The number of people who only play on a phone are fairly large.

As someone who used to play the Android port of wesnoth back in the day, I 100% agree, porting the newer versions to Android would be an excellent way to bring in more players without affecting existing ones.

>     * I'd like to remind that GUI2 is fully skinnable (that was one of the main reasons for switching to GUI2 in the first place). So it is not very hard to have a setting in the Preferences that allows one to switch between multiple themes. In fact, it should be possible for Add-ons to have their own themes as well. (Sadly, nobody managed to add that yet, but I do have it on my TODO list). We can ship Wesnoth with two themes: those who prefer the previous look could simply choose that theme, while those who prefer this one can select this. There can even be others. GIMP for example ships with 5 themes, LibreOffice with multiple themes as well. Though I can't seem to recall a game that does have multiple themes, I believe Wesnoth can fairly easily manage that, since moddablity is one of its core strengths.

Ok, making it toggleable would make it more bearable at least, I would appreciate that. (hopefully story text UI is included too and/or can be toggled seperately)

ForestDragon-wesnoth commented:
> Unfortunately, nostalgia can only keep a game on life support, never bring in new players. If we don't at least try to keep a healthy player count, I worry that Wesnoth will continue withering away as it has been.

Think of it from the other way around too - if specific changes lose more old players than bring new ones, the game dies even faster than if nothing was done. And lately controversial changes became far more common than in pre-1.16 days.

> While we're on the topic, I posted some in-game UI mockups some time ago on the discord. Not gorgeous, but I do feel something like this would be a lot more appealing for new players. Colors and styles would need changes to better match with the existing menus UI, if we're not planning on any global style changes.

To be blunt, this UI idea looks nothing like wesnoth anymore, and looks more like some kind of sci-fi RPG

Pentarctagon commented:
> Think of it from the other way around too - if specific changes lose more old players than bring new ones, the game dies even faster than if nothing was done. And lately controversial changes became far more common than in pre-1.16 days.

Still better to at least try rather than doing nothing.

ForestDragon-wesnoth commented:
> Still better to at least try rather than doing nothing.

So far this is what I described above, "chasing progress for the sake of progress". The dev team believes the current state of the game needs to be very radically changed, but it's the old state of the game that most of the playerbase originally loved. And the idea of attracting new players seems to have become misused as a mantra to justify any and all changes that old players dislike, whether a significant enough amount of new players actually like the changes or not.

Versions before 1.18 were generally a lot more moderate with what got changed each update, and most of the time the new versions truly felt like an net improvement. from 1.18/upcoming 1.19 content it has become a very mixed bag to say the least. Never before in wesnoth's update history have I seen so many controversial changes in such a short time.

babaissarkar commented:
I think #4531 is a better place to discuss/debate things than this PR, please.
Besides, I did provide an alternative which tries its best to cater to the needs of both old and new player alike. I'd rather not like to alienate any side if there is a solution that is feasible for both sides.

ForestDragon-wesnoth commented:
> I think #4531 is a better place to discuss/debate things than this PR, please. 
Oh, ok, sorry.

> Besides, I did provide an alternative which tries its best to cater to the needs of both old and new player alike. I'd rather not like to alienate any side if there is a solution that is feasible for both sides.

Thanks. Having the option to toggle between the two modes makes this a lot more better.

babaissarkar commented:
**GUI2 Theme Switching support**
1. It is now possible to change the `gui2_theme` key from **Display** tab of **Preferences** dialog
![Screenshot from 2024-06-22 18-47-58](https://github.com/wesnoth/wesnoth/assets/8469888/eb485e2c-1ac1-434c-80f0-e013e72bec55)
2. The `data/gui` folder is reorganized in a manner better suited to multiple themes. `_main.cfg` is modified accordingly.
![Screenshot from 2024-06-22 18-47-01](https://github.com/wesnoth/wesnoth/assets/8469888/95a70a01-75ad-40d6-aabc-78a858a1c353)

soliton- commented:
Having a preference for "Interface theme" and "UI theme" is confusing since they basically mean the same thing.

I suggest to rename "Interface theme" to something like "in-game theme" or however we call the UI when playing a scenario since that is basically what the GUI1 theme is and for what we eventually need an equivalent when it is converted to GUI2.

soliton- commented on src/gui/dialogs/title_screen.cpp:
```diff
@@ -208,7 +208,19 @@ void title_screen::init_callbacks()
+						 << "</span>"
+						 << tip.text().str().substr(id+1);
+				widget["label"] = tip_text.str();
+			}
```
This seems like it could use a comment.

babaissarkar replied:
explanatory comments added

babaissarkar commented on src/gui/dialogs/preferences_dialog.cpp:
```diff
@@ -159,6 +161,28 @@ void preferences_dialog::set_theme_list(menu_button& theme_list)
+			current_theme = i;
+		}
+		if (current_theme == 0) {
+			i++;
```
If multiple themes have same id, the first one is selected. It works if the first theme matches.

soliton- replied:
If the first theme matches `current_theme` gets set to 0 so `i` is incremented anyway so next match assigns another `i`. Maybe rather use a bool to remember a match.

babaissarkar replied:
what about making `current_theme` to have -1 at start, to identify no match?

babaissarkar replied:
used boolean for checking.

babaissarkar commented on src/game_config.cpp:
```diff
@@ -313,8 +314,13 @@ void load_config(const config &v)
+		game_title_background = backgrounds.at(r);
+
+		// Allow game_title to be empty
+		game_title            = i["game_title"].str().empty() ? game_title_background : i["game_title"].str();
```
More details? Do you mean checking with [isblank()](https://en.cppreference.com/w/cpp/string/byte/isblank)?

babaissarkar replied:
this line stops a warning which shows up if `game_title` is empty.

soliton- replied:
I mean https://github.com/wesnoth/wesnoth/blob/e67a91caf22429c56cd2012e324cd8ee212b5481/src/config_attribute_value.cpp#L357
It checks if the attribute was specified at all.

Nothing wrong with the warnings and maybe nothing wrong with ignoring them as well? I'm not suggesting that we do not set `game_title` just that it's possible to not set it. Whether that looks good or not is a different question.

Or is it fundamentally broken if `game_title` is empty?

babaissarkar replied:
> Nothing wrong with the warnings and maybe nothing wrong with ignoring them as well?
If it is possible for `game_title` to be empty/blank, then isn't that warning redundant? Maybe we can just remove that warning.
Going one step further, `game_title` is simply another image that gets blitted on top of `game_title_background`, maybe that's redundant as well, since we have the IPF `~BLIT()`.

soliton- replied:
> Going one step further, `game_title` is simply another image that gets blitted on top of `game_title_background`, maybe that's redundant as well, since we have the IPF `~BLIT()`.

Sounds like it. Though since you make it a list now then you'd have to repeat the BLIT each time? Though maybe you want to BLIT a bit differently for different images...

babaissarkar replied:
> Sounds like it. Though since you make it a list now then you'd have to repeat the BLIT each time? Though maybe you want to BLIT a bit differently for different images...

No, only one image is randomly selected from the list, so you're always blitting two images.

babaissarkar replied:
I think changing the condition on the warning instead of assigning a value here would be better. Instead of showing the warning when `game_title` is empty, show it instead when both `game_title` and `game_title_background` are empty (the background becomes black in this case.)

babaissarkar replied:
The WML part:
https://github.com/wesnoth/wesnoth/blob/e67a91caf22429c56cd2012e324cd8ee212b5481/data/gui/window/title_screen.cfg#L34-L50

babaissarkar replied:
Changed the condition for the warning, only happens when both game_title and game_background are empty.

babaissarkar commented on src/game_config.cpp:
```diff
@@ -313,8 +314,13 @@ void load_config(const config &v)
-		game_title_background = i["game_title_background"].str();
+		// Select a background at random
+		const auto backgrounds = utils::split(i["game_title_background"].str());
+		int r = rand() % (backgrounds.size());
```
Added checking.

babaissarkar commented:
> I suggest to rename "Interface theme" to something like "in-game theme" or however we call the UI when playing a scenario since that is basically what the GUI1 theme is and for what we eventually need an equivalent when it is converted to GUI2.

Renamed in latest commit to "In-game theme".

babaissarkar commented:
Changes in b6dbe6a712652ea684db1fc894ad5db5cde354e4: Preliminary theme loading from add-ons

If an add-on contains a `gui.cfg` file in its root with a valid theme, it will be loaded as well, and the user will be able to change to this theme from `UI Theme` list.

soliton- commented:
Would be good to put theme loading in a new PR. I was thinking to use `gui/_main.cfg` like in mainline.

babaissarkar commented:
> Would be good to put theme loading in a new PR. I was thinking to use `gui/_main.cfg` like in mainline.

Moved to #9057. Also changed the gui filename to `gui/_main.cfg` (inside addon folder)

babaissarkar commented:
Does it make sense for each theme to have its own set of background images? (Currently they are being specified by the game_config file)

soliton- commented:
Is there a requirement that a theme needs to use what is specified in game_config?

CelticMinstrel commented:
Thinking about it, the point of those images being in the game config is so that a custom core can replace them. So I think the title screen probably _should_ get its background images from the game config.

babaissarkar commented:
> Thinking about it, the point of those images being in the game config is so that a custom core can replace them. So I think the title screen probably _should_ get its background images from the game config and not from the chosen GUI theme.

That's valid, how about using the value in game_config as a fallback? If a theme specifies a background, then that is used instead. I'm asking about this because the classic theme pairs better with the map background while the new theme works better with the story backgrounds from TRoW. (plus this allows me to preserve the present theme's appearance).

CelticMinstrel commented:
> That's valid, how about using the value in `game_config` as a fallback? If a theme specifies a background, then that is used instead.

I think this is a tough question, because if you imagine a custom core that rethemes Wesnoth to be a completely different genre of game, then the theme's background definitely isn't going to fit with the custom core. I'm not sure what the best answer is.

demario-w commented:
First, let me make clear that if my tone sound harsh, I have no personal issue with the author of the quoted lines. lol

> > devs forget the fact that nostalgia is the main factor keeping the community active.
> 
> Unfortunately, nostalgia can only keep a game on life support, never bring in new players. If we don't at least try to keep a healthy player count, I worry that Wesnoth will continue withering away as it has been.

I don't see how nostalgia could be a bad thing. Obviously nostalgia only impacts the people who have a strong feeling with the game and its history. New players might not feel the nostalgia, but it doesn't mean it is not worth protecting for the people it means something.

Every commercial entertainment proposal understands that keeping a strong memory of its past and nourishing the feeling of historic foundations is essential to build a strong image even in the mind of new public. See how soccer's FIFA and american NFL are keeping referring to historic players and games to grow their brand. Some part of the new public might ignore it but some new recruits are attracted to it. Other clear examples are Disney animation or any movie franchise.

On the other hand, wesnoth is systematically erasing its past every 2 years: clearing the add-on server, introducing incompatibilities in the code, not supporting access to historic records of famous games, breaking stats compatibility for the most foolish reasons... As @ForestDragon-wesnoth says, this trend has been speeding-up since the last version.

> 
> IMO the biggest potential turn-offs for new players are UI, the tutorial/first campaign, and RNG - and thus I feel that those are probably the most important issues we should aim to address.
> 

The claim that Wesnoth should target a popular success is the paragon mistaken conviction from developers who discard Wesnoth history.

Wesnoth is not designed to be mainstream. The luck factor, the time it is required to play, the standard for art that prevents frequent update (see recent statement on AI art)... It is good that some barrier is clear at the entrance of the game to help the people who are not part of the natural public (strategy game, turn based, 2D pixel graphics...) see it at first glance and don't waste their time figuring out every aspect of the game before they realize that is not their cup of tea.

----

Finally I must call the developers out for not admitting how laughable it is to pretend to speak for _the people who are not playing yet_ (the "new players") to override the objections of the people who have spend years on this game and are voicing their concerns. It is just clearly another way to pass the message that the devs "know better" and they "are always right".

soliton- commented:
> Note that even without a key for background in a theme, its still possible to override game_config's value, since a theme can replace the entire titlescreen dialog.

So nothing to do IMO.

> But I cannot do the random background change in WML, it is implemented in C++. (Or can I?)

WFL does have a dice operator for randomness. So I think something like this should work:
`switch(1d6, 1, 'image1', 2, 'image2', 3, 'image3', 'default image')`

babaissarkar commented:
How does this look?
![Screenshot from 2024-07-12 23-05-15](https://github.com/user-attachments/assets/3d112083-06d7-437d-84db-686fc342b424)

(The help button is hovered, my screenshot tool decided not to capture the mouse)

babaissarkar commented:
The language button now shows the display name of the language in that language instead of just showing the language code. (e1ea5adfccaf7a2ac9e820d80bc81c399e328251)

![Screenshot from 2024-07-13 13-34-55](https://github.com/user-attachments/assets/240bcc29-9231-47c7-a872-9eb5035394dd)
![Screenshot from 2024-07-13 13-34-26](https://github.com/user-attachments/assets/ca7959e9-64a4-4393-95d0-bd0b2322961c)

babaissarkar commented:
@CelticMinstrel 

> I think the version number needs an outline for contrast.

Look a bit bad, any other ideas?
Maybe can just change the "About" button's caption to something like "v1.19"?

soliton- commented:
Other than the outline being cut off at both sides it's not that bad?

Vultraz reviewed (approved):
LGTM. We can come back and refine the design later, but for now, excellent work! A UI refresh off to a goof start!

(I especially like the use of the script font for the first letter of the tip of the day)
